### PR TITLE
Data science

### PR DESCRIPTION
**Pull request by yuvalroth**
Data science - take 1

### CODE REVIEW

File: data_science/lib/data_manager.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+class DataManager
+  @sample_size = 0
+  attr_accessor :reader
+  def load(reader)
+    @reader = reader
+  end
+
+  def sample_size(cohort_name = nil)
+    raise ArgumentError, "reader not initialized" if @reader.nil?
```
Comment (zsyed1) [inline data_science/lib/data_manager.rb]: You can do `unless @reader`

File: data_science/specs/data_loader_spec.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
+require 'rspec'
+require_relative 'constants'
+require_relative '../lib/json_reader'
+
+describe 'JSON Reader' do
+  describe 'Initialization' do
+    before :each do
+      @reader = JsonReader.new
```
Comment (zsyed1) [inline data_science/specs/data_loader_spec.rb]: Use `let` instead:
```
let(:reader) { JsonReader.new }
```

File: data_science/specs/data_manager_spec.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
+require 'rspec'
+require_relative '../lib/data_manager'
+require_relative '../lib/json_reader'
+require_relative 'constants'
+
+describe 'Data Manager' do
+  describe 'Initialization' do
+    before :each do
+      @reader = JsonReader.new
```
Comment (zsyed1) [inline data_science/specs/data_manager_spec.rb]: Use `let` instead - that's the convention here :smile:

File: data_science/lib/data_manager.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    total = sample_size(cohort)
+    ((converted.to_f / total.to_f) * 100).round(2)
+  end
+
+  def calculate_chi_square
+    values = {}
+    prepare_data_by_group(values, :agroup, "A")
+    prepare_data_by_group(values, :bgroup, "B")
+    tester = ABAnalyzer::ABTest.new values
+    # if this number is smaller than 0.05 we are good
```
Comment (zsyed1) [inline data_science/lib/data_manager.rb]: Ruby community doesn't like comments ... I disagree with the Ruby community :smiley:

File: data_science/specs/data_manager_spec.rb [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      @reader.load_data(JSON_FILE_PATH)
+      @data_manager = DataManager.new
+      @data_manager.load(@reader)
+    end
+
+    it 'should have 0.32 rate' do
+      expect(@data_manager.calculate_chi_square).to eq(0.032)
+    end
+  end
+end
```
Comment (zsyed1) [inline data_science/specs/data_manager_spec.rb]: Nice work! This looks done to me. Please change the `before` blocks to `let` statements though. Not sure if this needs resubmittal @djkotowski

File: data_science/lib/data_manager.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  @sample_size = 0
+  attr_accessor :reader
+  def load(reader)
+    @reader = reader
+  end
+
+  def sample_size(cohort_name = nil)
+    raise ArgumentError, "reader not initialized" if @reader.nil?
+    if cohort_name.nil?
+      @reader.data_hash.count
```
Comment (djkotowski) [inline data_science/lib/data_manager.rb]: You can use a guard clause to get rid of the multi-line if-statement here.

```ruby
return @reader.data_hash.count unless cohort_name
@reader.data_hash.count { |item| item['cohort'] == cohort_name }
```

File: data_science/lib/data_manager.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    else
+      @reader.data_hash.count { |item| (item['cohort']) == cohort_name }
```
Comment (djkotowski) [inline data_science/lib/data_manager.rb]: Redundant parantheses around `item['cohort']`

File: data_science/lib/data_manager.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    end
+  end
+
+  def conversion_size(cohort_name = nil)
+    raise ArgumentError, "reader not initialized" if @reader.nil?
+    if cohort_name.nil?
+      @reader.data_hash.count { |item| (item['result']) == 1 }
+    else
+      cohort_conversion_size(cohort_name)
+    end
```
Comment (djkotowski) [inline data_science/lib/data_manager.rb]: Here's another place where guard clauses will help.

File: data_science/lib/data_manager.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  def cohort_conversion_size(name)
+    @reader.data_hash.count do |item|
+      (item['cohort']) == name && (item['result'] == 1)
```
Comment (djkotowski) [inline data_science/lib/data_manager.rb]: Redundant parentheses around `item['cohort']` and the second half of the `&&` expression.

File: data_science/lib/data_manager.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    end
+  end
+  private :cohort_conversion_size
```
Comment (djkotowski) [inline data_science/lib/data_manager.rb]: Prefer using `private` as a separator between public and private methods.

Example from wikibooks:
```ruby
class Example
    def methodA
    end
    
    private # all methods that follow will be made private: not accessible for outside objects
    
    def methodP
    end
  end
```

File: data_science/lib/data_manager.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    converted = conversion_size(cohort)
+    total = sample_size(cohort)
+    ((converted.to_f / total.to_f) * 100).round(2)
+  end
+
+  def calculate_chi_square
+    values = {}
+    prepare_data_by_group(values, :agroup, "A")
+    prepare_data_by_group(values, :bgroup, "B")
+    tester = ABAnalyzer::ABTest.new values
```
Comment (djkotowski) [inline data_science/lib/data_manager.rb]: Wrap args with parentheses.

File: data_science/lib/data_manager.rb (hunk 4 above)
Comment (djkotowski) [inline data_science/lib/data_manager.rb]: Though `0.05` may change depending on the desired confidence level, thus why comments lie.

File: data_science/specs/data_loader_spec.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    end
+    context 'when initializing data reader' do
+      it 'should not be nil' do
+        expect(@reader).to_not be_nil
+      end
+    end
+  end
+
+  describe '.load_data' do
```
Comment (djkotowski) [inline data_science/specs/data_loader_spec.rb]: Prefix instance methods with `#` and class methods with `.`

File: data_science/specs/data_loader_spec.rb [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    context 'when loading data' do
+      it 'should load all elements' do
+        expect(JsonReader.new.load_data(JSON_FILE_PATH).count).to eq(2892)
```
Comment (djkotowski) [inline data_science/specs/data_loader_spec.rb]: Use the reader defined above.

File: data_science/specs/data_manager_spec.rb [hunk 14]
```diff
@@ -0,0 +1,125 @@
+require 'rspec'
+require_relative '../lib/data_manager'
+require_relative '../lib/json_reader'
+require_relative 'constants'
+
+describe 'Data Manager' do
+  describe 'Initialization' do
```
Comment (djkotowski) [inline data_science/specs/data_manager_spec.rb]: Use `describe "#new"` here.

File: data_science/specs/data_manager_spec.rb [hunk 15]
```diff
@@ -0,0 +1,125 @@
+require 'rspec'
+require_relative '../lib/data_manager'
+require_relative '../lib/json_reader'
+require_relative 'constants'
+
+describe 'Data Manager' do
```
Comment (djkotowski) [inline data_science/specs/data_manager_spec.rb]: Describe the class itself.

```ruby
describe JsonReader do
```

File: data_science/specs/data_manager_spec.rb [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+  describe 'Initialization' do
+    before :each do
+      @reader = JsonReader.new
+      @data_manager = DataManager.new
+    end
+
+    context 'when initialized data manager' do
```
Comment (djkotowski) [inline data_science/specs/data_manager_spec.rb]: Get rid of `data manager` here.

File: data_science/specs/data_loader_spec.rb [hunk 17]
```diff
@@ -0,0 +1,24 @@
+require 'rspec'
+require_relative 'constants'
+require_relative '../lib/json_reader'
+
+describe 'JSON Reader' do
```
Comment (djkotowski) [inline data_science/specs/data_loader_spec.rb]:
```ruby
describe JsonReader
```

File: data_science/specs/data_manager_spec.rb (hunk 3 above)
Comment (djkotowski) [inline data_science/specs/data_manager_spec.rb]: Mock out `JsonReader`.  By including it in this spec, you're testing `DataManager` **and** `JsonReader`.

See [RSpec Mocks](https://www.relishapp.com/rspec/rspec-mocks/docs).

**Comment by djkotowski** [comment]
Nice iteration.  Take a look at the comments, refactor and resubmit when ready.

**Pull request closed without merging**